src/partition/repart.c

```diff
@@ -523,35 +523,40 @@ static uint64_t free_area_available(const FreeArea *a) {
         return a->size - a->allocated;
 }
 
-static uint64_t free_area_available_for_new_partitions(Context *context, const FreeArea *a) {
-        uint64_t avail;
-
+static uint64_t free_area_current_end(Context *context, const FreeArea *a) {
         assert(context);
         assert(a);
+        assert(a->after);
+        assert(a->after->offset != UINT64_MAX);
+        assert(a->after->current_size != UINT64_MAX);
 
-        /* Similar to free_area_available(), but takes into account that the required size and padding of the
-         * preceding partition is honoured. */
-
-        avail = free_area_available(a);
-        if (a->after) {
-                uint64_t need, space_end, new_end;
+        /* Calculate where the free area ends, based on the offset of the partition preceding it. */
+        return round_up_size(a->after->offset + a->after->current_size, context->grain_size) + free_area_available(a);
+}
```

Comment:
if this is now a helper of its own, maybe also add a code path dealing with the first free area, i.e. where `a->after` is NULL. In that case you can just return `free_area_available()` after all the offset is zero...